lib/base/io-engine.cpp

```diff
@@ -16,60 +16,73 @@
 
 using namespace icinga;
 
-CpuBoundWork::CpuBoundWork(boost::asio::yield_context yc)
+CpuBoundWork::CpuBoundWork(boost::asio::yield_context yc, boost::asio::io_context::strand& strand)
 	: m_Done(false)
 {
 	auto& ioEngine (IoEngine::Get());
+	auto& sem (ioEngine.m_CpuBoundSemaphore);
+	std::unique_lock<std::mutex> lock (sem.Mutex);
 
-	for (;;) {
-		auto availableSlots (ioEngine.m_CpuBoundSemaphore.fetch_sub(1));
+	if (sem.FreeSlots) {
+		--sem.FreeSlots;
+		return;
+	}
+
+	auto cv (Shared<AsioConditionVariable>::Make(ioEngine.GetIoContext()));
+	bool gotSlot = false;
+	auto pos (sem.Waiting.insert(sem.Waiting.end(), IoEngine::CpuBoundQueueItem{&strand, cv, &gotSlot}));
 
-		if (availableSlots < 1) {
-			ioEngine.m_CpuBoundSemaphore.fetch_add(1);
-			IoEngine::YieldCurrentCoroutine(yc);
-			continue;
+	lock.unlock();
+
+	try {
+		cv->Wait(yc);
+	} catch (...) {
+		std::unique_lock<std::mutex> lock (sem.Mutex);
+
+		if (gotSlot) {
```

Comment:
You can just use `pos->GotSlot` instead here and don't need to keep track of a bool type.

Reply:
Items get moved out of sem.Waiting which invalidates _pos_. _gotSlot_ tells me whether _pos_ is still valid or not.

Reply:
cppreference says this:
> Adding, removing and moving the elements within the list or across several lists does not invalidate the iterators or references. An iterator is invalidated only when the corresponding element is deleted.

I would simply use a pointer to `CpuBoundQueueItem` for the queue instead then.
```cxx
auto item = IoEngine::CpuBoundQueueItem{&strand, cv, false};
auto pos (sem.Waiting.emplace(sem.Waiting.end(), &item));
```

Reply:
> only when the corresponding element is deleted.

Exactly that gotSlot tells.